### PR TITLE
✨Add support for dynamically computing active semester

### PR DESCRIPTION
**Pull request by vikaspotluri123**
refs #22 

Might be easier to review on a commit-by-commit basis

### CODE REVIEW

File: lib/services/current-semester.js [hunk 1]
```diff
@@ -1,4 +1,126 @@
...
+/** @type {SemesterActiveFunction} */
+const isFallActive = (currentMonth, currentDay, currentYear) => {
+	const isActiveAugustDay = currentMonth === AUGUST && currentDay >= 1;
+	const isActiveDecemberDay = currentMonth === DECEMBER && currentDay <= 28;
+	const isActive = isActiveAugustDay || isActiveDecemberDay || (currentMonth > AUGUST && currentMonth < DECEMBER);
+
+	return isActive ? `${currentYear}F` : null;
+};
+
+// Fall is active between December 7 and January 25
```
Comment (joshcosta) [inline lib/services/current-semester.js:42]: winter

**Comment by joshcosta** [comment]
The logic looks good. It needs tests.

How will this be exposed to client?

**Comment by joshcosta** [comment]
I think we should be more lenient with the active semesters to allow people to add classes during registration if they choose. 

Proposed changes:
- Spring: November 15 to May 25
- Summer: April 1 to August 25
- Fall: April 1 to December 25
- Winter: November 15 to January 25

**Comment by vikaspotluri123** [comment]
By active do you mean allowed? We might need to rethink how we do course creation, with the inclusion of a semester dropdown when there are multiple allowed semesters 🤔 

I think we should expose semester data in the core-data endpoint since we should only refresh active semesters daily, and the app suspends before that

**Comment by vikaspotluri123** [comment]
But to respond to the dates, IMO that's way too early. @ramsays what do you think?

File: lib/services/current-semester.js (hunk 1 above)
Comment (vikaspotluri123) [reply lib/services/current-semester.js]: fixed

**Comment by vikaspotluri123** [comment]
@joshcosta @ramsays once this gets merged, I'd like to abstract the current semester logic to gradebook/utils since it can probably be used in other repos

**Comment by ramsays** [comment]
Correct me if I'm wrong, but we are using "active" in two different meanings here, one to say whether a semester is available to be created/edited (which matches the client restrictions?) vs the semester that is the dashboard default.

I don't mind the original dates that Vikas set for create/editable, given that most people don't fix their schedule/get their syllabi until pretty close to semester start. So I don't think we need to have such wide options.

We probably should clear up our terminology as stated above.

**Comment by vikaspotluri123** [comment]
> Correct me if I'm wrong, but we are using "active" in two different meanings here, one to say whether a semester is available to be created/edited (which matches the client restrictions?) vs the semester that is the dashboard default.

`activeSemester` is the semester that's default - so dashboard, import (unless we decide to change the logic), etc.

`validSemesters` are those that can be edited - e.g. CRUD data

Where do you see `activeSemester` being used in the context of `validSemesters`? If there's any confusion in the codebase, it's definitely something that should be fixed!

File: lib/services/current-semester.js [hunk 2]
```diff
@@ -1,5 +1,127 @@
...
+	const currentDay = currentDate.getDate();
+
+	module.exports.activeSemester = _getActiveSemester(currentMonth, currentDay, currentYear);
+	module.exports.validSemesters = [
+		isWinterActive(currentMonth, currentDay, currentYear),
+		isSpringActive(currentMonth, currentDay, currentYear),
+		isSummerActive(currentMonth, currentDay, currentYear),
+		isFallActive(currentMonth, currentDay, currentYear)
+	].filter(Boolean);
+}
```
Comment (ramsays) [inline lib/services/current-semester.js:123]: Combined use of "active" and "valid"!

**Pull request merged**